keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -242,22 +242,102 @@ type AdmissionRequest struct {
+run may have opinions on those new structures (like setting an `imagePullPolicy` on all containers). 
 
-Mutations from in-tree plugins will not trigger this process as they are well-ordered but if 
-there is any mutation by webhooks, all of the plugins including in-tree ones will be run again.
+To allow mutating admission plugins to observe changes made by other plugins, regardless of order,
+admission will add the ability to re-run mutating plugins (including webhooks) a single time if 
+there is any change made by a mutating webhook on the first pass.
 
-This feature would be would be opt in and defaulted to false for `v1beta1`.
+1. Initial mutating admission pass
+2. If any webhook plugins returned patches modifying the object, do a second mutating admission pass
+   1. Run all in-tree mutating admission plugins
+   2. Run all applicable webhook mutating admission plugins that indicate they want to be reinvoked
 
-The API representation and behavior for this feature is still under design and will be updated/approved here prior to implementation.
+Mutating plugins that are reinvoked must be idempotent, able to successfully process an object they have already 
+admitted and potentially modified. This will be clearly documented in admission webhook documentation,
+examples, and test guidelines. Mutating admission webhooks *should* already support this (since any 
+change they can make in an object could already exist in the user-provided object), and any webhooks
+that do not are broken for some set of user input.
+
+Note that idempotence (whether a webhook can successfully operate on its own output) is distinct from
+the `sideEffects` indicator, which describes whether a webhook can safely be called in `dryRun` mode.
+
+The reinvocation behavior will be opt-in, so `reinvocationPolicy` defaults to `"Never"`.
+
+```golang
+type Webhook struct {
+    ...
+    // reinvocationPolicy indicates whether this webhook should be called multiple times as part of a single admission evaluation.
+    // Allowed values are "Never" and "IfNeeded".
+    //
+    // Never: the webhook will not be called more than once in a single admission evaluation.
+    //
+    // IfNeeded: the webhook will be called at least one additional time as part of the admission evaluation
+    // if the object being admitted is modified by other admission plugins after the initial webhook call.
+    // Webhooks that specify this option *must* be idempotent, able to process objects they previously admitted.
+    // Note:
+    // * the number of additional invocations is not guaranteed to be exactly one.
+    // * if additional invocations result in further modifications to the object, webhooks are not guaranteed to be invoked again.
+    // * webhooks that use this option may be reordered to minimize the number of additional invocations.
+    // * to validate an object after all mutations are guaranteed complete, use a validating admission webhook instead.
```

Comment:
Add something about how webhooks with side effects can / should avoid double-actuating those side effects.

Will validating webhooks get called multiple times, or is this change only for mutating webhooks? If the latter, we should make sure this option doesn't appear in the documentation for validating webhooks.

Reply:
> Add something about how webhooks with side effects can / should avoid double-actuating those side effects.

Updated with a clarification and recommendation to use this for webhooks with side-effects.

> Will validating webhooks get called multiple times, or is this change only for mutating webhooks? If the latter, we should make sure this option doesn't appear in the documentation for validating webhooks.

Reinvocation would not apply for validating webhooks. There's not a good way to document a struct differently for different uses, but we can clarify that in the docs for this option, and make sure validation doesn't permit selecting this option for validating webhooks.

Reply:
If this type is shared, then I think we need to split it in two.

Reply:
There's significant shared code between mutating and validating webhooks, and splitting the type would use up the time remaining before freeze this week. Is a validating webhook with a reinvocation policy of `Never` objectionable?

Reply:
Yes, because if we add that field to validating webhooks, we can't get rid of it in the future.

Reply:
split types in the doc

Reply:
Here's the split PR: https://github.com/kubernetes/kubernetes/pull/78491
Moved pretty fast on this one so extra eyes from reviewers are (extra) welcome!